### DISCUSSION TITLE
Runtime plugin selection via command flags

### DISCUSSION THREAD

**Discussion by SamHasler** [oclif/oclif#1478] [Q&A] [answered]
I'd like to provide several different implementations for some functionality in different plugins (which would allow users also to configure their own plugins to provide it if they wished). 

I think from looking at how hooks work it seems the best way to do this will be to have a custom  "get_providers" hook that each plugin implements and have the hooks return a name, and a function that the command can use. The command would then have to be given the name of a provider to use in a flag.

Is that going to be the best way to allow runtime plugin selection or is there another pattern I could follow?

**Comment by mdonnalley** [comment, marked as answer]
Oclif doesn't have anything builtin for that but your approach sounds reasonable to me

**Reply by SamHasler** [reply]
Thank you for the confirmation. I've got something working. I probably don't need to pass the name back as it also returns the plugin so I can just look at the plugin name and/or alias.
